topdown/builtins/builtins.go

```diff
@@ -28,6 +29,67 @@ func (c Cache) Get(k interface{}) (interface{}, bool) {
 	return v, ok
 }
 
+// We use an ast.Object for the cached keys/values because a naive
+// map[ast.Value]ast.Value will not correctly detect value equality of
+// the member keys.
```

Comment:
Golang has strict comparison semantics: https://go.dev/ref/spec#Comparison_operators

As best I could tell, Go's default strict comparison semantics do not work for raw `ast.Value`s, because we need custom `Compare` and `Equal` functions to properly do value comparisons on those types. :slightly_frowning_face:

When I was using the naive 2-level `map[string]map[ast.Value]ast.Value`, I found that value-identical `ast.Value`s would be treated as *different* keys, likely due to their pointers being different.  I then switched to `map[string]ast.Object`, because `ast.Object` has the appropriate logic to handle `ast.Value` types internally for its keys/values, and that seems to work exactly as intended.